Initial cert validation test

benjaminleonard opened the pull request:
Fixes #2580 and related to #2578 

Idea from @augustuswm to soft validate the cert. Unfortunately I think we need a library to parse the cert but I lazy imported it to avoid adding to the main bundle since most users will not see this.

Principally the risk here is what I dont know about certs:

1. wildcard handling might need to be sophisticated
2. certs that arrange slightly different in a way that `parseCertificate()` doesn't catch

But feels, like the image soft validation, that it could be a good quality of life improvement for something that is potentially very painful when not done correctly. We might want to label the `CN` and `SAN` on the messages list of found names.

<img width="499" alt="image" src="https://github.com/user-attachments/assets/aa26a784-c898-4c94-afe4-e7167b3d6b73">

benjaminleonard committed:
Initial cert validation test

github-actions[bot] committed:
Bot commit: format with prettier

benjaminleonard committed:
Spacing tweak

benjaminleonard commented on app/components/form/fields/TlsCertsField.tsx:
```diff
@@ -147,3 +176,88 @@ const AddCertModal = ({ onDismiss, onSubmit, allNames }: AddCertModalProps) => {
+  if (patternParts.length !== domainParts.length) return false
+
+  return patternParts.every(
+    (part, i) => part === '*' || part.toLowerCase() === domainParts[i].toLowerCase()
```
Would it just be the first part that accepts a wildcard?

benjaminleonard replied:
I presume it would be the left-most wildcard subdomain is the only allowable?

Would need adjusting, and some unit tests also.

augustuswm replied:
Correct only the left most subdomain can be a wildcard. For example `*.r3.oxide-preview.com`  does not match `silo.sys.r3.oxide-preview.com`

benjaminleonard committed:
Merge branch 'tls-cert-soft-validation' of https://github.com/oxidecomputer/console into tls-cert-soft-validation

david-crespo commented on package.json:
```diff
@@ -51,10 +52,12 @@
     "match-sorter": "^6.3.4",
     "md5": "^2.3.0",
     "mousetrap": "^1.6.5",
+    "npm": "^10.9.1",
```
lol you `npm i npm i`

benjaminleonard replied:
haha what an idiot

benjaminleonard committed:
oop

benjaminleonard committed:
Wildcard and error improvements with tests

benjaminleonard commented:
Added tests, improved wildcard handling and also a message in case the certificate cannot be parsed.

<img width="498" alt="image" src="https://github.com/user-attachments/assets/ba61204a-c785-4bb1-857f-13a0b2617de3">

Message copy needs work.

benjaminleonard committed:
More pattern matching tweaks

charliepark commented:
Is this your card? https://github.com/oxidecomputer/console/blob/require_tls_cert/app/forms/idp/util.ts#L16-L17

david-crespo committed:
move dynamic import to avoid suspense/lazy thing

benjaminleonard committed:
Merge branch 'main' into tls-cert-soft-validation

augustuswm commented:
This looks awesome, thanks for picking it up. I think initial rack install will certainly benefit from it.

benjaminleonard commented:
> Is this your card? https://github.com/oxidecomputer/console/blob/require_tls_cert/app/forms/idp/util.ts#L16-L17

Do we need some copy to caveat incase that is not correct. I remember some talk about a proxy perhaps throwing it off?

david-crespo commented on app/components/form/fields/TlsCertsField.tsx:
```diff
@@ -103,10 +114,18 @@ type AddCertModalProps = {
+  const { data: certValidation } = useQuery({
+    queryKey: ['validateImage', ...(file ? [file.name, file.size, file.lastModified] : [])],
+    queryFn: file ? () => file.text().then(parseCertificate) : skipToken,
+  })
```
@charliepark pointed out we might want to set the stale time directly here. I'm going to experiment with that and also maybe tweaking the query key.

david-crespo committed:
merge main and resolve conflicts

david-crespo committed:
improve copy, don't opt out of validation if there's no silo name

david-crespo commented:
Tweaked the copy and fixed a bug where it would only do validation if the silo name was non-empty in the parent form. I changed it so it only opts out of the part of the validation that requires the silo name.

<img width="443" height="259" alt="image" src="https://github.com/user-attachments/assets/e882a729-af31-4597-9489-34b68d363a04" />

benjaminleonard committed:
Merge branch 'main' into tls-cert-soft-validation

benjaminleonard committed:
Use getDelegatedDomain helper instead of hardcoded domain

The hardcoded r2.oxide-preview.com produces a false "domain mismatch"
warning on every rack that isn't r2 (dogfood, colo, customer installs).
Reuse the same helper the IdP create form uses to derive the suffix
from window.location.

benjaminleonard committed:
Address review feedback on cert validation

- Rename stale 'validateImage' query key to 'validateCert'
- Drop redundant *.sys.<domain> match arm: it's subsumed by
  expectedDomain matching (a *.sys.<domain> SAN already matches
  <silo>.sys.<domain>) and only the early return guards the
  empty-siloName case anyway
- Rename commonName to commonNames (it's a string[])
- Spread certValidation ?? {} so the undefined-spread isn't subtle
- Tighten "Could not be parsed" copy

benjaminleonard committed:
Tidy comments and remove dead || [] fallbacks

getField('CN') and .map() both already return string[]. Replace the
hand-wavy bare-'*' comment with the RFC 6125 reference.

benjaminleonard committed:
Tighten matchesDomain: case, trailing dot, '*.' edge case

- Normalize case and strip trailing dots on both pattern and domain up
  front (RFC 6125 §6.4 case-insensitive, RFC 4592 trailing-dot FQDN).
  Wildcard branch was previously case-sensitive.
- Reject pathological '*.' which would have matched any 2-label domain
  via endsWith('') being true.

benjaminleonard committed:
parseCertificate accepts PEM chains and surrounding whitespace

Users frequently paste a chain (leaf + intermediates). The previous
single-PEM parse would throw and report "Couldn't parse" for a valid
file. Extract the first PEM block (the leaf, by convention) and parse
that. Side benefit: tolerates leading whitespace / BOM.

benjaminleonard committed:
Warn when uploaded cert is already expired

parseCertificate now returns notAfter; CertDomainNotice renders a
'notice'-variant message when the cert is past its expiry. Expiry and
domain-mismatch warnings can stack — they're independent. Factored the
duplicated docs-link block into a SiloCertsDocsLink helper.

benjaminleonard committed:
E2E coverage for cert parse/expiry/mismatch notices

- Assert "Couldn't parse certificate" appears after the existing garbage
  chooseFile upload
- Swap the second cert upload to a real (expired, non-matching) PEM and
  assert both the expiry and mismatch notices render, plus that the
  parse-failure notice is gone

benjaminleonard committed:
E2E: upload a matching cert and assert no notice fires

Before the bad-PEM step, upload a self-signed cert whose SANs cover
other-silo.sys.placeholder and isn't expired. Confirms the
soft-validation notices are absent in the happy path, not just present
in the unhappy one.

benjaminleonard commented:
Updated from main and a little review:

**Major**
- Replaced `r2.oxide-preview.com` with `getDelegatedDomain(window.location)` as per @charliepark's snippet

**Smaller polish**
- Renamed the stale `validateImage` query key, dropped the redundant match arm
- `matchesDomain` now normalizes case (wildcard branch was case-sensitive) and trailing dots up front, and rejects the pathological `*.` pattern.
- Warns when the uploaded cert is already past its `notAfter`. Stacks with the domain-mismatch warning when both apply.

**Tests**
- Vitest: wildcard + uppercase, trailing dot, `*.` rejection, multi-PEM bundle, whitespace/BOM, expiry parse.
- Playwright (`silos.e2e.ts`): asserts "Couldn't parse certificate" on the existing garbage upload, then uploads a valid matching cert (no notice fires) and a real expired non-matching PEM (both notices fire), exercising the FileField → useQuery → parseCertificate → matchesDomain → notice pipeline end to end.

---

Can we get one last sanity check and revisit merging?

david-crespo committed:
merge main and resolve conflicts

david-crespo committed:
fix wildcard regex for label boundary

david-crespo committed:
prefer SANs for cert domain matching

david-crespo committed:
don't need the import to be dynamic

david-crespo commented:
I took out the dynamic import in 765210b796207700ccce3633cffbac405401a98e because even though technically certs aren't required by the API in the silo create body, in practice they are necessarily pretty much all the time. So that import will always happen, might as well absorb it into the chunk for the form. Had the robot compare before/after:

<img width="610" height="194" alt="image" src="https://github.com/user-attachments/assets/03d6297c-3a78-452b-b2ac-de2babecfc72" />

david-crespo committed:
one last small cleanup

benjaminleonard commented:
LGTM

<!-- linear:replyTo:2504016213 -->

david-crespo commented:
Let's give this this a last look but I think it's ready to go.
